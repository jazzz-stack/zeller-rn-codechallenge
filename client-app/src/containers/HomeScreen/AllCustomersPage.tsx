import React from 'react';
import { FlatList, KeyboardAvoidingView, Platform } from 'react-native';
import { UserCard, LoadingState, EmptyState } from '../../components/Molecules';
import { ZellerCustomer } from '../../types/User';
import { styles } from './HomeScreenStyle';

import { RefreshControlProps } from 'react-native';

interface AllCustomersPageProps {
  customers: ZellerCustomer[];
  loading: boolean;
  error: string | null;
  refreshing: boolean;
  onRefresh: () => void;
  onEditCustomer: (user: ZellerCustomer) => void;
  onDeleteCustomer: (user: ZellerCustomer) => void;
  onScroll: (event: any) => void;
  onEndReached: () => void;
  refreshControl?: React.ReactElement<RefreshControlProps>;
}

export const AllCustomersPage: React.FC<AllCustomersPageProps> = ({
  customers,
  loading,
  error,
  refreshing,
  onRefresh,
  onEditCustomer,
  onDeleteCustomer,
  onScroll,
  onEndReached,
  refreshControl,
}) => (
  <KeyboardAvoidingView
    behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
    style={{ flex: 1 }}
  >
    <FlatList
      data={customers}
      keyExtractor={(item: ZellerCustomer) => item.id}
      renderItem={({ item }) => (
        <UserCard user={item} onEdit={onEditCustomer} onDelete={onDeleteCustomer} />
      )}
      ListEmptyComponent={loading && !error ? <LoadingState /> : <EmptyState />}
      {...(refreshControl ? { refreshControl } : {})}
      contentContainerStyle={[styles.listContent]}
      showsVerticalScrollIndicator={false}
      onScroll={onScroll}
      scrollEventThrottle={16}
      onEndReached={onEndReached}
      onEndReachedThreshold={0.5}
      initialNumToRender={10}
      maxToRenderPerBatch={10}
      refreshing={refreshing}
      onRefresh={onRefresh}
    />
  </KeyboardAvoidingView>
);
