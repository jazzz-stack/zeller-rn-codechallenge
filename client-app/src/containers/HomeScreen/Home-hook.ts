import {useEffect, useRef, useState, useMemo} from 'react';
import {Animated} from 'react-native';
import {useAppDispatch, useAppSelector} from '../../store/hooks';
import {
  loadCustomersFromLocal,
  syncCustomersFromAPI,
  backgroundSyncCustomersFromAPI,
  deleteCustomer,
  setSearchQuery,
  setSelectedFilter,
} from '../../store/slices/customersSlice';
import {ZellerCustomer, UserRole} from '../../types/User';
import {HomeScreenProps} from '../../navigation/schema_types/navigation-schema';

export interface CustomerSection {
  title: string;
  data: ZellerCustomer[];
}

export const useHomeHook = (navigation: HomeScreenProps['navigation']) => {
  const dispatch = useAppDispatch();

  // Local UI state
  const [isSearching, setIsSearching] = useState(false);
  const [refreshing, setRefreshing] = useState(false);
  const [isFocused, setIsFocused] = useState(false);
  
  // Delete modal state
  const [showDeleteModal, setShowDeleteModal] = useState(false);
  const [customerToDelete, setCustomerToDelete] = useState<ZellerCustomer | null>(null);
  const [isDeleting, setIsDeleting] = useState(false);
  
  // Animation for header visibility
  const headerTranslateY = useRef(new Animated.Value(0)).current;
  const lastScrollY = useRef(0);
  const [isHeaderVisible, setIsHeaderVisible] = useState(true);

  // Redux state
  const {
    filteredCustomers: customers,
    loading,
    error,
    searchQuery,
    selectedFilter,
  } = useAppSelector(state => state.customers);

  // Create sorted sections for customers
  const customerSections = useMemo(() => {
    if (!customers || customers.length === 0) {
      return [];
    }

    // Sort customers alphabetically by name
    const sortedCustomers = [...customers].sort((a, b) =>
      a.name.toLowerCase().localeCompare(b.name.toLowerCase())
    );

    // Group customers by first letter
    const grouped = sortedCustomers.reduce((acc: Record<string, ZellerCustomer[]>, customer) => {
      const firstLetter = customer.name.charAt(0).toUpperCase();
      if (!acc[firstLetter]) {
        acc[firstLetter] = [];
      }
      acc[firstLetter].push(customer);
      return acc;
    }, {});

    // Convert to sections array and sort alphabetically
    return Object.keys(grouped)
      .sort()
      .map(letter => ({
        title: letter,
        data: grouped[letter],
      }));
  }, [customers]);

  // Smart loading strategy on mount
  useEffect(() => {
    const initializeData = async () => {
      try {
        // Step 1: Always load from local database first
        const localResult = await dispatch(loadCustomersFromLocal()).unwrap();
        // Step 2: Determine sync strategy based on local data availability
        if (localResult.length > 0) {
          // User has local data - use background sync (won't show errors in UI)
          dispatch(backgroundSyncCustomersFromAPI());
        } else {
          // First time user - use regular sync (will show errors in UI if network fails)
          try {
            await dispatch(syncCustomersFromAPI()).unwrap();
          } catch (syncError) {
            // Error will be shown in UI via Redux state
          }
        }
      } catch (loadError) {
        // If loading local data fails, try API sync anyway
        try {
          await dispatch(syncCustomersFromAPI()).unwrap();
        } catch (syncError) {
          // Error will be shown in UI via Redux state
        }
      }
    };
    
    initializeData();
  }, [dispatch]);

  // Business logic handlers
  const handleSearchToggle = (_shouldSearch: boolean) => {
    setIsSearching(!isSearching);
    handleSearchChange(''); // Clear search when toggling
  };

  const handleAddCustomer = () => {
    navigation.navigate('AddCustomer');
  };

  const handleEditCustomer = (user: ZellerCustomer) => {
    navigation.navigate('EditCustomer', {customer: user});
  };

  const handleDeleteCustomer = (user: ZellerCustomer) => {
    setCustomerToDelete(user);
    setShowDeleteModal(true);
  };

  const confirmDeleteCustomer = async () => {
    if (!customerToDelete) return;
    
    setIsDeleting(true);
    try {
      await dispatch(deleteCustomer(customerToDelete.id)).unwrap();
      // Success - modal state will be handled by the component
    } catch (err) {
      // Error - modal state will be handled by the component  
    } finally {
      setIsDeleting(false);
      setShowDeleteModal(false);
      setCustomerToDelete(null);
    }
  };

  const cancelDeleteCustomer = () => {
    setShowDeleteModal(false);
    setCustomerToDelete(null);
    setIsDeleting(false);
  };

  const handleRefresh = async () => {
    dispatch(loadCustomersFromLocal());
  };

  const handlePullToRefresh = async () => {
    setRefreshing(true);
    try {
      // Step 1: Always load local data first to show something immediately
      const localResult = await dispatch(loadCustomersFromLocal()).unwrap();
      dispatch(backgroundSyncCustomersFromAPI());

    } catch (error) {
      // If local loading fails, try background sync anyway
      dispatch(backgroundSyncCustomersFromAPI());
    } finally {
      // Ensure refreshing state is always cleared
      setRefreshing(false);
    }
  };

  const handleSearchChange = (query: string) => {
    dispatch(setSearchQuery(query));
  };

  const handleFilterChange = (filter: UserRole) => {
    dispatch(setSelectedFilter(filter));
  };

  const handleSearchOpen = () => {
    setIsSearching(true);
    dispatch(setSelectedFilter('ALL'));
  };

  const handleSyncFromAPI = async () => {
    try {
      await dispatch(syncCustomersFromAPI()).unwrap();
      // If sync succeeds, error state will be automatically cleared by Redux
    } catch (error) {
      // Error will be shown in UI via Redux state
      // Don't log here to prevent duplicate console logs
    }
  };

  // Handle scroll to show/hide header
  const handleScroll = (event: any) => {
    const currentScrollY = event.nativeEvent.contentOffset.y;
    const diff = currentScrollY - lastScrollY.current;
    // Only trigger animation if scroll difference is significant
    if (Math.abs(diff) > 10) {
      if (diff > 0 && currentScrollY > 30 && isHeaderVisible) {
        // Scrolling down - hide header
        setIsHeaderVisible(false);
        Animated.timing(headerTranslateY, {
          toValue: -150, // Move header completely out of view
          duration: 300,
          useNativeDriver: true,
        }).start();
      } else if (diff < 0 && !isHeaderVisible) {
        // Scrolling up - show header
        setIsHeaderVisible(true);
        Animated.timing(headerTranslateY, {
          toValue: 0,
          duration: 300,
          useNativeDriver: true,
        }).start();
      }
      lastScrollY.current = currentScrollY;
    }
  };

  return {
    customers,
    customerSections,
    loading,
    error,
    searchQuery,
    selectedFilter,
    isSearching,
    refreshing,
    headerTranslateY,
    isHeaderVisible,
    showDeleteModal,
    customerToDelete,
    isDeleting,
    isFocused,
    handleSearchToggle,
    handleAddCustomer,
    handleEditCustomer,
    handleDeleteCustomer,
    confirmDeleteCustomer,
    cancelDeleteCustomer,
    handleRefresh,
    handlePullToRefresh,
    handleSearchChange,
    handleFilterChange,
    handleSearchOpen,
    handleSyncFromAPI,
    handleScroll,
    setIsFocused,
  };
};
