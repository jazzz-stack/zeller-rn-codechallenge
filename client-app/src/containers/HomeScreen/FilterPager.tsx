import React from 'react';
import PagerView from 'react-native-pager-view';
import { View, RefreshControl } from 'react-native';
import { AllCustomersPage } from './AllCustomersPage';
import { AdminCustomersPage } from './AdminCustomersPage';
import { ManagerCustomersPage } from './ManagerCustomersPage';
import { ZellerCustomer, UserRole } from '../../types/User';

const FILTERS: UserRole[] = ['ALL', 'Admin', 'Manager'];

export interface FilterPagerProps {
  customers: ZellerCustomer[];
  loading: boolean;
  error: string | null;
  refreshing: boolean;
  customerSections: any[];
  selectedFilter: UserRole;
  onRefresh: () => void;
  onFilterChange: (filter: UserRole) => void;
  onEditCustomer: (user: ZellerCustomer) => void;
  onDeleteCustomer: (user: ZellerCustomer) => void;
  onScroll: (event: any) => void;
}

export const FilterPager: React.FC<FilterPagerProps> = ({
  customers,
  loading,
  error,
  refreshing,
  customerSections,
  selectedFilter,
  onRefresh,
  onFilterChange,
  onEditCustomer,
  onDeleteCustomer,
  onScroll,
}) => {
  const pagerRef = React.useRef<PagerView>(null);
  const selectedIndex = FILTERS.indexOf(selectedFilter);

  React.useEffect(() => {
    if (pagerRef.current) {
      pagerRef.current.setPage(selectedIndex);
    }
  }, [selectedIndex]);

  // Pagination state for each filter
  const [pageMap, setPageMap] = React.useState({ ALL: 1, Admin: 1, Manager: 1 });
  const PAGE_SIZE = 20;

  const getFilteredCustomers = (filter: UserRole) => {
    if (filter === 'ALL') return customers;
    return customers.filter(c => c.role === filter);
  };

  const getPaginatedCustomers = (filter: UserRole) => {
    const all = getFilteredCustomers(filter);
    return all.slice(0, pageMap[filter] * PAGE_SIZE);
  };

  const handleEndReached = (filter: UserRole) => {
    const all = getFilteredCustomers(filter);
    if (getPaginatedCustomers(filter).length < all.length) {
      setPageMap(prev => ({ ...prev, [filter]: prev[filter] + 1 }));
    }
  };

  const refreshControl = (
    <RefreshControl
      refreshing={refreshing}
      onRefresh={onRefresh}
      colors={["#3498DB"]}
      tintColor="#3498DB"
      title="Pull to refresh..."
      progressViewOffset={100}
      progressBackgroundColor="#fff"
    />
  );

  // Track last page to avoid unnecessary setPage calls
  const lastPage = React.useRef(selectedIndex);

  React.useEffect(() => {
    if (pagerRef.current && selectedIndex !== lastPage.current) {
      pagerRef.current.setPage(selectedIndex);
      lastPage.current = selectedIndex;
    }
  }, [selectedIndex]);

  return (
    <PagerView
      style={{ flex: 1 }}
      ref={pagerRef}
      testID="pager-view"
      onPageSelected={e => {
        const idx = e.nativeEvent.position;
        if (idx !== selectedIndex) {
          onFilterChange(FILTERS[idx]);
        }
      }}
    >
      <View key="ALL" style={{ flex: 1 }}>
        <AllCustomersPage
          customers={getPaginatedCustomers('ALL')}
          loading={loading}
          error={error}
          refreshing={refreshing}
          onRefresh={onRefresh}
          onEditCustomer={onEditCustomer}
          onDeleteCustomer={onDeleteCustomer}
          onScroll={onScroll}
          onEndReached={() => handleEndReached('ALL')}
          refreshControl={refreshControl}
        />
      </View>
      <View key="Admin" style={{ flex: 1 }}>
        <AdminCustomersPage
          customers={getPaginatedCustomers('Admin')}
          loading={loading}
          error={error}
          refreshing={refreshing}
          onRefresh={onRefresh}
          onEditCustomer={onEditCustomer}
          onDeleteCustomer={onDeleteCustomer}
          onScroll={onScroll}
          onEndReached={() => handleEndReached('Admin')}
          refreshControl={refreshControl}
        />
      </View>
      <View key="Manager" style={{ flex: 1 }}>
        <ManagerCustomersPage
          customers={getPaginatedCustomers('Manager')}
          loading={loading}
          error={error}
          refreshing={refreshing}
          onRefresh={onRefresh}
          onEditCustomer={onEditCustomer}
          onDeleteCustomer={onDeleteCustomer}
          onScroll={onScroll}
          onEndReached={() => handleEndReached('Manager')}
          refreshControl={refreshControl}
        />
      </View>
    </PagerView>
  );
};
