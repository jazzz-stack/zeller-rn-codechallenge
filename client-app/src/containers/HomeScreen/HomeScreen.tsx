import React from 'react';
import {
  View,
  SafeAreaView,
  Text,
  Animated,
  Platform,
} from 'react-native';
import { FilterPager } from './FilterPager';
import {FloatingButton} from '../../components/Atoms';
import {
  UserCard,
  LoadingState,
  EmptyState,
  ErrorState,
  CustomerFilterHeader,
  ConfirmationModal,
} from '../../components/Molecules';
import {ZellerCustomer} from '../../types/User';
import {styles} from './HomeScreenStyle';
import {HomeScreenProps} from '../../navigation/schema_types/navigation-schema';
import {useHomeHook} from './Home-hook';

const HomeScreen = ({navigation}: HomeScreenProps) => {
  const {
    customers,
    customerSections,
    loading,
    error,
    searchQuery,
    selectedFilter,
    isSearching,
    refreshing,
    headerTranslateY,
    showDeleteModal,
    customerToDelete,
    isDeleting,
    isFocused,
    handleSearchToggle,
    handleAddCustomer,
    handleEditCustomer,
    handleDeleteCustomer,
    confirmDeleteCustomer,
    cancelDeleteCustomer,
    handlePullToRefresh,
    handleSearchChange,
    handleFilterChange,
    handleSearchOpen,
    handleScroll,
    handleSyncFromAPI,
    setIsFocused,
  } = useHomeHook(navigation);

  return (
    <SafeAreaView style={[styles.safeArea]}>
      {error && customerSections?.length === 0 ? (
        <ErrorState
          error={error}
          onRetry={handleSyncFromAPI}
          loading={loading}
        />
      ) : (
        <View style={[styles.container]}>
          <Animated.View
            style={[
              styles.animatedHeader,
              { transform: [{ translateY: headerTranslateY }] },
            ]}
          >
            <CustomerFilterHeader
              isSearching={isSearching}
              searchQuery={searchQuery}
              selectedFilter={selectedFilter}
              onSearchChange={handleSearchChange}
              onSearchToggle={handleSearchToggle}
              onFilterChange={handleFilterChange}
              onSearchOpen={handleSearchOpen}
              onFocus={setIsFocused}
            />
            {customers.length > 0 && (
              <Text style={styles.title}>
                {'Zeller Customers'}
                {` (${customers.length})`}
              </Text>
            )}
          </Animated.View>
          <FilterPager
            customers={customers}
            loading={loading}
            error={error}
            refreshing={refreshing}
            customerSections={customerSections}
            selectedFilter={selectedFilter}
            onRefresh={handlePullToRefresh}
            onFilterChange={handleFilterChange}
            onEditCustomer={handleEditCustomer}
            onDeleteCustomer={handleDeleteCustomer}
            onScroll={handleScroll}
          />
          <FloatingButton
            onPress={handleAddCustomer}
            variant="primary"
            size="medium"
          />
        </View>
      )}

      {/* Confirmation Modal */}
      <ConfirmationModal
        visible={showDeleteModal}
        type="delete"
        message={`Are you sure you want to delete ${customerToDelete?.name}?`}
        onConfirm={confirmDeleteCustomer}
        onCancel={cancelDeleteCustomer}
        isLoading={isDeleting}
        confirmText="Delete"
      />
    </SafeAreaView>
  );
};

export default HomeScreen;
