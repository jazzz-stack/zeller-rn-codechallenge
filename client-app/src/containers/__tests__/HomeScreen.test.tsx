import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react-native';
import { Alert, Animated, Platform } from 'react-native';
import HomeScreen from '../HomeScreen/HomeScreen';
import { useHomeHook } from '../HomeScreen/Home-hook';

// Mock the custom hook
jest.mock('../HomeScreen/Home-hook', () => ({
  useHomeHook: jest.fn(),
}));

// Mock all the components
jest.mock('../../components/Atoms', () => ({
  FloatingButton: ({ onPress, variant, size }: any) => {
    const React = require('react');
    const { TouchableOpacity, Text } = require('react-native');
    
    return React.createElement(TouchableOpacity, { 
      testID: 'floating-button',
      onPress: onPress
    }, React.createElement(Text, {}, `Floating ${variant} ${size}`));
  },
}));

jest.mock('../../components/Molecules', () => ({
  UserCard: ({ user, onEdit, onDelete }: any) => {
    const React = require('react');
    const { View, Text, TouchableOpacity } = require('react-native');
    
    return React.createElement(View, { testID: `user-card-${user.id}` },
      React.createElement(Text, { testID: 'user-name' }, user.name),
      React.createElement(Text, { testID: 'user-email' }, user.email),
      React.createElement(Text, { testID: 'user-role' }, user.role),
      React.createElement(TouchableOpacity, { 
        testID: 'edit-button',
        onPress: () => onEdit(user)
      }, React.createElement(Text, {}, 'Edit')),
      React.createElement(TouchableOpacity, { 
        testID: 'delete-button',
        onPress: () => onDelete(user.id)
      }, React.createElement(Text, {}, 'Delete'))
    );
  },
  LoadingState: () => {
    const React = require('react');
    const { View, Text } = require('react-native');
    
    return React.createElement(View, { testID: 'loading-state' },
      React.createElement(Text, {}, 'Loading...')
    );
  },
  EmptyState: () => {
    const React = require('react');
    const { View, Text } = require('react-native');
    
    return React.createElement(View, { testID: 'empty-state' },
      React.createElement(Text, {}, 'No customers found')
    );
  },
  ErrorState: ({ error, onRetry }: any) => {
    const React = require('react');
    const { View, Text, TouchableOpacity } = require('react-native');
    
    return React.createElement(View, { testID: 'error-state' },
      React.createElement(Text, { testID: 'error-message' }, error),
      React.createElement(TouchableOpacity, { 
        testID: 'retry-button',
        onPress: onRetry
      }, React.createElement(Text, {}, 'Retry'))
    );
  },
  CustomerFilterHeader: ({ 
    isSearching, 
    searchQuery, 
    selectedFilter, 
    onSearchChange, 
    onSearchToggle,
    onFilterChange,
    onSearchOpen 
  }: any) => {
    const React = require('react');
    const { View, Text, TouchableOpacity, TextInput } = require('react-native');
    
    return React.createElement(View, { testID: 'filter-header' },
      React.createElement(TouchableOpacity, { 
        testID: 'search-toggle',
        onPress: onSearchToggle
      }, React.createElement(Text, {}, isSearching ? 'Close Search' : 'Open Search')),
      isSearching && React.createElement(TextInput, {
        testID: 'search-input',
        value: searchQuery,
        onChangeText: onSearchChange,
        placeholder: 'Search customers...'
      }),
      React.createElement(Text, { testID: 'selected-filter' }, selectedFilter),
      React.createElement(TouchableOpacity, { 
        testID: 'filter-button',
        onPress: () => onFilterChange('Admin')
      }, React.createElement(Text, {}, 'Filter Admin'))
    );
  },
  ConfirmationModal: ({ visible, message, onConfirm, onCancel, isLoading }: any) => {
    const React = require('react');
    const { Modal, View, Text, TouchableOpacity } = require('react-native');
    
    if (!visible) return null;
    
    return React.createElement(Modal, { 
      testID: 'delete-confirmation-modal',
      visible: visible 
    },
      React.createElement(View, { testID: 'modal-content' },
        React.createElement(Text, { testID: 'modal-title' }, 'Delete Customer'),
        React.createElement(Text, { testID: 'modal-message' }, message),
        React.createElement(TouchableOpacity, {
          testID: 'modal-cancel-button',
          onPress: onCancel,
          disabled: isLoading
        }, React.createElement(Text, {}, 'Cancel')),
        React.createElement(TouchableOpacity, {
          testID: 'modal-confirm-button',
          onPress: onConfirm,
          disabled: isLoading
        }, React.createElement(Text, {}, isLoading ? 'Deleting...' : 'Delete'))
      )
    );
  },
}));

// Mock Alert
jest.spyOn(Alert, 'alert');

describe('HomeScreen', () => {
  const mockNavigation = {
    navigate: jest.fn(),
    goBack: jest.fn(),
  };

  const mockProps = {
    navigation: mockNavigation as any,
    route: {} as any,
  };

  const mockCustomers = [
    {
      id: '1',
      name: 'John Doe',
      email: 'john@example.com',
      role: 'Admin' as const,
    },
    {
      id: '2',
      name: 'Jane Smith',
      email: 'jane@example.com',
      role: 'Manager' as const,
    },
  ];

  const mockCustomerSections = [
    {
      title: 'Admin',
      data: [mockCustomers[0]],
    },
    {
      title: 'Manager',
      data: [mockCustomers[1]],
    },
  ];

  const mockHookReturnValue = {
    customers: mockCustomers,
    customerSections: mockCustomerSections,
    loading: false,
    error: null,
    searchQuery: '',
    selectedFilter: 'ALL' as const,
    isSearching: false,
    refreshing: false,
    headerTranslateY: new Animated.Value(0),
    isHeaderVisible: true,
    
    showDeleteModal: false,
    customerToDelete: null,
    isDeleting: false,
    
    handleSearchToggle: jest.fn(),
    handleAddCustomer: jest.fn(),
    handleEditCustomer: jest.fn(),
    handleDeleteCustomer: jest.fn(),
    confirmDeleteCustomer: jest.fn(),
    cancelDeleteCustomer: jest.fn(),
    handleRefresh: jest.fn(),
    handlePullToRefresh: jest.fn(),
    handleSearchChange: jest.fn(),
    handleFilterChange: jest.fn(),
    handleSearchOpen: jest.fn(),
    handleSyncFromAPI: jest.fn(),
    handleScroll: jest.fn(),
    setIsFocused: jest.fn(),
    isFocused: false,
  };

  beforeEach(() => {
    jest.clearAllMocks();
    (useHomeHook as jest.Mock).mockReturnValue(mockHookReturnValue);
  });

  it('renders HomeScreen with customer list', () => {
    render(<HomeScreen {...mockProps} />);
    
    expect(screen.getAllByTestId('user-card-1').length).toBeGreaterThan(0);
    expect(screen.getAllByTestId('user-card-2').length).toBeGreaterThan(0);
    expect(screen.getAllByText('John Doe').length).toBeGreaterThan(0);
    expect(screen.getAllByText('Jane Smith').length).toBeGreaterThan(0);
    expect(screen.getByTestId('floating-button')).toBeTruthy();
    expect(screen.getByTestId('filter-header')).toBeTruthy();
  });

  it('renders loading state when loading', () => {
    (useHomeHook as jest.Mock).mockReturnValue({
      ...mockHookReturnValue,
      loading: true,
      customers: [],
      customerSections: [],
    });

    render(<HomeScreen {...mockProps} />);
    
    expect(screen.getAllByTestId('loading-state').length).toBeGreaterThan(0);
    expect(screen.getAllByText('Loading...').length).toBeGreaterThan(0);
  });

  it('renders empty state when no customers', () => {
    (useHomeHook as jest.Mock).mockReturnValue({
      ...mockHookReturnValue,
      loading: false,
      customers: [],
      customerSections: [],
    });

    render(<HomeScreen {...mockProps} />);
    
    expect(screen.getAllByTestId('empty-state').length).toBeGreaterThan(0);
    expect(screen.getAllByText('No customers found').length).toBeGreaterThan(0);
  });

  it('renders error state when there is an error', () => {
    const mockError = 'Network error';
    (useHomeHook as jest.Mock).mockReturnValue({
      ...mockHookReturnValue,
      error: mockError,
      customers: [], // No customers when error occurs
      customerSections: [], // No sections when error occurs
    });

    render(<HomeScreen {...mockProps} />);
    
    expect(screen.getByTestId('error-state')).toBeTruthy();
    expect(screen.getByText(mockError)).toBeTruthy();
    expect(screen.getByTestId('retry-button')).toBeTruthy();
  });

  it('handles floating button press', () => {
    render(<HomeScreen {...mockProps} />);
    
    const floatingButton = screen.getByTestId('floating-button');
    fireEvent.press(floatingButton);
    
    expect(mockHookReturnValue.handleAddCustomer).toHaveBeenCalledTimes(1);
  });

  it('handles user edit action', () => {
    render(<HomeScreen {...mockProps} />);
    
    const editButton = screen.getAllByTestId('edit-button')[0];
    fireEvent.press(editButton);
    
    expect(mockHookReturnValue.handleEditCustomer).toHaveBeenCalledWith(mockCustomers[0]);
  });

  it('handles user delete action', () => {
    render(<HomeScreen {...mockProps} />);
    
    const deleteButton = screen.getAllByTestId('delete-button')[0];
    fireEvent.press(deleteButton);
    
    expect(mockHookReturnValue.handleDeleteCustomer).toHaveBeenCalledWith('1');
  });

  it('handles search toggle', () => {
    render(<HomeScreen {...mockProps} />);
    
    const searchToggle = screen.getByTestId('search-toggle');
    fireEvent.press(searchToggle);
    
    expect(mockHookReturnValue.handleSearchToggle).toHaveBeenCalledTimes(1);
  });

  it('handles search when searching is active', () => {
    (useHomeHook as jest.Mock).mockReturnValue({
      ...mockHookReturnValue,
      isSearching: true,
      searchQuery: 'test',
    });

    render(<HomeScreen {...mockProps} />);
    
    const searchInput = screen.getByTestId('search-input');
    expect(searchInput).toBeTruthy();
    expect(screen.getByDisplayValue('test')).toBeTruthy();
    
    fireEvent.changeText(searchInput, 'new search');
    expect(mockHookReturnValue.handleSearchChange).toHaveBeenCalledWith('new search');
  });

  it('handles filter change', () => {
    render(<HomeScreen {...mockProps} />);
    
    const filterButton = screen.getByTestId('filter-button');
    fireEvent.press(filterButton);
    
    expect(mockHookReturnValue.handleFilterChange).toHaveBeenCalledWith('Admin');
  });

  it('handles retry when in error state', () => {
    (useHomeHook as jest.Mock).mockReturnValue({
      ...mockHookReturnValue,
      error: 'Network error',
      customers: [], // No customers when error occurs
      customerSections: [], // No sections when error occurs
    });

    render(<HomeScreen {...mockProps} />);
    
    const retryButton = screen.getByTestId('retry-button');
    fireEvent.press(retryButton);
    
    expect(mockHookReturnValue.handleSyncFromAPI).toHaveBeenCalledTimes(1);
  });

  it('displays customer count in title when customers exist', () => {
    render(<HomeScreen {...mockProps} />);
    
    expect(screen.getByText('Zeller Customers (2)')).toBeTruthy();
  });

  it('displays selected filter in header', () => {
    (useHomeHook as jest.Mock).mockReturnValue({
      ...mockHookReturnValue,
      selectedFilter: 'Manager',
      customers: [], // Remove customers to avoid text conflicts
      customerSections: [],
    });

    render(<HomeScreen {...mockProps} />);
    
    const selectedFilter = screen.getByTestId('selected-filter');
    expect(selectedFilter).toBeTruthy();
    expect(selectedFilter.props.children).toBe('Manager');
  });

  it('renders section headers correctly', () => {
    render(<HomeScreen {...mockProps} />);
    
    expect(mockHookReturnValue.customerSections).toHaveLength(2);
    expect(mockHookReturnValue.customerSections[0].title).toBe('Admin');
    expect(mockHookReturnValue.customerSections[1].title).toBe('Manager');
  });

  it('calls useHomeHook with navigation', () => {
    render(<HomeScreen {...mockProps} />);
    
    expect(useHomeHook).toHaveBeenCalledWith(mockNavigation);
  });

  it('handles refreshing state', () => {
    const mockRefreshingHookValue = {
      ...mockHookReturnValue,
      refreshing: true,
      showDeleteModal: false,
      customerToDelete: null,
      isDeleting: false,
      setIsFocused: jest.fn(),
      isFocused: false,
      confirmDeleteCustomer: jest.fn(),
      cancelDeleteCustomer: jest.fn(),
    };
    
    (useHomeHook as jest.Mock).mockReturnValue(mockRefreshingHookValue);

    render(<HomeScreen {...mockProps} />);
    
    expect(mockRefreshingHookValue.refreshing).toBe(true);
  });

  // Snapshot Tests
  it('matches snapshot when error', () => {
    (useHomeHook as jest.Mock).mockReturnValue({
      ...mockHookReturnValue,
      error: 'Network error',
      customers: [], // No customers when error occurs
      customerSections: [], // No sections when error occurs
    });

    const component = render(<HomeScreen {...mockProps} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  // Platform-specific tests to improve branch coverage
  describe('Platform-specific behavior', () => {
    it('should render correctly on iOS platform', () => {
      const originalPlatform = Platform.OS;
      Platform.OS = 'ios';

      (useHomeHook as jest.Mock).mockReturnValue({
        ...mockHookReturnValue,
        isFocused: false,
      });

      render(<HomeScreen {...mockProps} />);
      
      expect(screen.getByTestId('floating-button')).toBeTruthy();
      
      Platform.OS = originalPlatform;
    });

    it('should render correctly on Android platform', () => {
      const originalPlatform = Platform.OS;
      Platform.OS = 'android';

      (useHomeHook as jest.Mock).mockReturnValue({
        ...mockHookReturnValue,
        isFocused: false,
      });

      render(<HomeScreen {...mockProps} />);
      
      expect(screen.getByTestId('floating-button')).toBeTruthy();
      
      Platform.OS = originalPlatform;
    });

    it('should apply focus-based padding on Android when isFocused is true', () => {
      const originalPlatform = Platform.OS;
      Platform.OS = 'android';

      (useHomeHook as jest.Mock).mockReturnValue({
        ...mockHookReturnValue,
        isFocused: true,
      });

      render(<HomeScreen {...mockProps} />);
      
      expect(screen.getByTestId('floating-button')).toBeTruthy();
      
      Platform.OS = originalPlatform;
    });

    it('should not apply focus-based padding on Android when isFocused is false', () => {
      const originalPlatform = Platform.OS;
      Platform.OS = 'android';

      (useHomeHook as jest.Mock).mockReturnValue({
        ...mockHookReturnValue,
        isFocused: false,
      });

      render(<HomeScreen {...mockProps} />);
      
      expect(screen.getByTestId('floating-button')).toBeTruthy();
      
      Platform.OS = originalPlatform;
    });

    it('should apply focus-based padding on Android when both isFocused and isSearching are true', () => {
      const originalPlatform = Platform.OS;
      Platform.OS = 'android';

      (useHomeHook as jest.Mock).mockReturnValue({
        ...mockHookReturnValue,
        isFocused: true,
        isSearching: true,
      });

      render(<HomeScreen {...mockProps} />);
      
      expect(screen.getByTestId('floating-button')).toBeTruthy();
      
      Platform.OS = originalPlatform;
    });
  });
});
