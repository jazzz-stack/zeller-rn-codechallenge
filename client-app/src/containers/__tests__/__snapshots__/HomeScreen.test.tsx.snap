// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen matches snapshot when error 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "backgroundColor": "#f8f9fa",
        "flex": 1,
      },
    ]
  }
>
  <View
    testID="error-state"
  >
    <Text
      testID="error-message"
    >
      Network error
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="retry-button"
    >
      <Text>
        Retry
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
